Sync `v2-3-stable` with `v2-3-test` to release `2.3.1`

ephraimbuddy opened the pull request:
Time for `2.3.1rc1`!

potiuk committed:
Remove custom signal handling in Triggerer (#23274)

There is a bug in CPython (fixed in March 2022 but not yet released) that
makes async.io handle SIGTERM improperly by using async unsafe
functions and hanging the triggerer receive SIGPIPE while handling
SIGTERN/SIGINT and deadlocking itself. Until the bug is handled
we should rather rely on standard handling of the signals rather than
adding our own signal handlers. Seems that even if our signal handler
just run exit(0) - it caused a race condition that led to the hanging.

More details:
   * https://bugs.python.org/issue39622
   * https://github.com/python/cpython/issues/83803

Fixes: #19260
(cherry picked from commit 6bdbed6c43df3c5473b168a75c50e0139cc13e88)

treyyi committed:
Fix typos in README.md and airflow_doc_issue_report.yml (#23294)

(cherry picked from commit c26796e31a9543cd8b45b50264128ac17455002c)

potiuk committed:
Fix empty image preparation (#23304)

Empty image preparation failed in CI because it was impossible to
build an empty image without buildkit. This change sets DOCKER_BUILDKIT
variable for empty image build which make it always use the buildkit.

(cherry picked from commit 4f6fe727a1af609595523bec4d1e6943f5e9075f)

bbovenzi committed:
Use <Time /> in Mapped Instance table (#23313)

The start/end dates for a mapped instance weren't updating when a user changed their timezone. Using <Time /> fixes that

(cherry picked from commit 98ec8c6990347fda60cbad33db915dc21497b1f0)

jhtimmins committed:
Add tags inside try block. (#21784)

(cherry picked from commit 190e911c46900459ab8d211acc76cefa18090cbf)

potiuk committed:
Unify style of communication with the users for Breeze. (#23311)

Fixes: #22906
(cherry picked from commit 505af06303d8160c71f6a7abe4792746f640083d)

bbovenzi committed:
Don't show grid actions if server would reject with permission denied (#23332)

* Add edit permission check for grid actions

* Remove if wrapper for meta tag

* Use dag.can_edit

(cherry picked from commit 67e8bdd691ded0e88760b3d59b9c58ba43fa9453)

bbovenzi committed:
Use run_id for ti.mark_success_url (#23330)

(cherry picked from commit a5f9df5c35645eeebb980e9288cd42dc7ac2cc54)

ephraimbuddy committed:
Update Airflow Release Doc (#23322)

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>
Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
(cherry picked from commit 2dfccfa82f602ab71db417bd9b45e86988ce29e4)

blag committed:
Fix update user auth stats (#23314)

(cherry picked from commit 1c886eaf99d9044cdf90bf7900dc489a4d40db79)

potiuk committed:
Remove confusion about upgrade-to-newer-dependencies breeze param (#23334)

The "upgrade-to-newer-dependencies" in the image can take "false"
"true" and <RANDOM> value (the RANDOM value is used to always
invalidate docker cache).

This has been carried to Breeze command line, but this was a source
of major confusion as the name of the parameter suggest bool value.

The change modifies the parameter to be flag, and when the flag
is set the parameter is set to random during image building.

That should help with recent bug when image was always rebuilt
without checking if it should be rebuilt.

(cherry picked from commit b6db0e90aeb30133086716a433cab9dca7408a54)

Bowrna committed:
resolving conflict (#23052)

(cherry picked from commit 94c3203e86252ed120d624a70aed571b57083ea4)

bbovenzi committed:
Hide some task instance attributes (#23338)

(cherry picked from commit f5f9c582a61e29a3bc007cb02a15579a42505565)

potiuk committed:
Cleaner default output when breeze starts (#23341)

There was a bit of noise printed when Breeze started:

* information about branch/python/image/backend used
* information about actions performed (like fixing permissions)
* information that docke image build is not needed
* warnings about missing variables

This PR marks all the messages as "info" and only prints them
when --verbose flag is used and it adds default values for the
variables that generated warnings.

(cherry picked from commit dd7002d905c89776c5d92fabd0591c02f2fc19f3)

jedcunningham committed:
Fix connection test button (#23345)

The connection test button was always disabled if any of your hooks had
import errors, for example because of a missing module. This handles
that scenario.

(cherry picked from commit f197030cea860351da07894879f647fe76c5751e)

bbovenzi committed:
Store grid view selection in url params (#23290)

* Add url params for dag_run_id and task_id

* Persist other search params

* simplify useSelection

* delete extra params

* remove API change

(cherry picked from commit 4c1fcee6ba1559fcb53836ac7fcccd68b1891d93)

potiuk committed:
Fix regeneration of breeze screenshots (#23344)

(cherry picked from commit c1528f71b055db3cf27ea4597f9550171fcb2678)

bbovenzi committed:
Fix broken task instance link in xcom list (#23367)

* Fix broken task instance link in xcom list

Add execution date back to the xcom list to be able to pass to the `task_instance_link()` function.
Long term, we should swap out the execution_date param for run_id

* Make execution date a search column

(cherry picked from commit 3920e6019347f70807252b3163ed017084ce2a92)

potiuk committed:
Fix mssql in the new Breeze (#23368)

The new Breeze did not use conditionally debian version to run
MsSQL docker compose.

This PR fixes it

(cherry picked from commit 9e11d208dac0951d4f3a5beda0d8801f4e70cb0a)

potiuk committed:
When exec fails in breeze we do not print stack-trace (#23342)

When you run exec and breeze is not running, there was a stack
trace printed rather than straightforward error message.

This fixes it - stacktrace is only printed now when verbose is
used. If not just error message is printed.

(cherry picked from commit e5ac9d8d0bb1c6b6ac7570d572acff2e86b8a29d)

potiuk committed:
Add missing --for-production parameter for new breeze docs building (#23376)

(cherry picked from commit da91d28c1846bed5d40f7c12e4b314c5f7beb812)

josh-fell committed:
Add doc notes for keyword-only args for `expand()` and `partial()` (#23373)

(cherry picked from commit d3028e1e9036a3c67ec4477eee6cd203c12f7f5c)

RNHTTR committed:
fix cli `airflow dags show` for mapped operator (#23339)

(cherry picked from commit 59e93106d55881163a93dac4a5289df1ba6e1db5)

jedcunningham committed:
Fix duplicated Kubernetes DeprecationWarnings (#23302)

(cherry picked from commit 45aadd267a68ecdc4f8491c79db26fa84024c5df)

tirkarthi committed:
Override pool for TaskInstance when pool is passed from cli. (#23258)

(cherry picked from commit 3970ea386d5e0a371143ad1e69b897fd1262842d)

potiuk committed:
Fix deriving of PyPI branch from airflow version (#23380)

The regexp expression for deriving right branch from version was
missing . for version numbers.

(cherry picked from commit 46da268567bd3a7197961db7f99bbe36f9708e7d)

ephraimbuddy committed:
Bump version to 2.3.0

jedcunningham committed:
Fix 2.3.0 release date in release notes

ephraimbuddy committed:
Update missing `version_added` in config.yml (#23387)

(cherry picked from commit 9a410e8da6bdbad537fadeda1849c45cdca073f3)

jedcunningham committed:
Add missing steps to release process (#23384)

(cherry picked from commit d670e571e499ad8e97dedfed3d81ed59469bce7c)

jedcunningham committed:
Fix `version_added` for `[sensors] default_timeout` (#23388)

(cherry picked from commit 62216112c6764381160793961ffc182eb77e466c)

potiuk committed:
Clarify 2.3.0 kubernetes min version is about library not cluster (#23398)

It was not clear from the release notes that the minimum version
of the kubernetes was about the library rather than cluster version.

Also README was not updated with min versions of Kubernetes cluster
for 2.3.0 version.

(cherry picked from commit 48d0267bdf6ae1a250d1f924847fe07726945bea)

ephraimbuddy committed:
Remove tagging of `constraint-x-y` branch from release process (#23399)

(cherry picked from commit 9faffc6c610adb86e421f24c24d1f8749551b286)

ephraimbuddy committed:
Ignore some files/directory when releasing source code (#23325)

(cherry picked from commit 16e1170f3b094dcc0b265856360283f6e645530a)

potiuk committed:
Improve handling of entry and exit to common Breeze commands (#23395)

This PR improves handling of both entry and exit to common
Breeze commands:

* at entry all common commands check if rebuild of image is needed
* when you exit and there is an error from shell commands, rather
  than printing stack trace an error message is printed

(cherry picked from commit 9a0080c20bb2c4a9c0f6ccf1ece79bde895688ac)

potiuk committed:
Fix "breeze-legacy" after building images was removed (#23404)

The `breeze-legacy` stopped working after building images were
removed as few parameters were still checked for allowed values
but they were missing,

This PR fixes it by removing the parameters.

(cherry picked from commit 8622808aa79531bcaa5099d26fbaf54b4afe931a)

potiuk committed:
Mark image as refreshed when pulled on CI (#23410)

One of the recent changes in Breeze (#23395) caused unnecessary
rebuilding of image when "build-docs" is run. On CI we build
image once and reuse it. However in case of Buld docs we missed
information that the image is "fresh" and we started rebuilding
it. This change marks the image as "refreshed" when it is pulled
with `--tag-as-latest` flag (which happens in CI).

(cherry picked from commit ce4a60ce279f691f428aa5f671180407e722ccbf)

bbovenzi committed:
Improve react www tests (#23329)

* Add shared test wrapper & treeData placeholder

* remove console log

* move testUtils to /utils

* change unnamed export

(cherry picked from commit 8f6b8551c325ff64f42080391a0f97b3a5d2e7d5)

bbovenzi committed:
Refactor code references from tree to grid (#23254)

(cherry picked from commit 22b49d334ef0008be7bd3d8481b55b8ab5d71c80)

c-thiel committed:
Add backward compatibility for core__sql_alchemy_conn__cmd (#23441)

(cherry picked from commit 0cdd401cda61006a42afba243f1ad813315934d4)

potiuk committed:
Improve verbose output of Breeze (#23446)

When you add --verbose or --dry-run options to breeze it will
print the commands it is executing (or is supposed to in dry-run
mode). The output contains environment variables as they
often contain crucial information to execute the command (for
example in docker-compose run commands it contains COMPOSE_FILE
variable which is the list of compose files that are used). This
is done in a fashion that you can copy the whole command and
execute it, but it very unfriendly for visual inspection as
all the variables were printed in one line and in semi-random
order and also the variables contained often all system variables
set by the shell before.

This change keeps the proerty of "we can copy&paste the command
and run it" but it improves the visual aspect of it:

1) each env variable is kept in one line
2) first all system variables are printed and then variables that
   were specifically added for this command
3) variables in each group are sorted alphabetically which helps
   in finding the variable you are looking for when you visually
   inspect the output.

(cherry picked from commit 6f146e721c81e9304bf7c0af66fc3d203d902dab)

raphaelauv committed:
[FIX] remove python 3.6 (#23409)

(cherry picked from commit af5463098ae30206c053dc96481dab48368cd4aa)

jedcunningham committed:
Fix `check_files.py` to work on new minor releases (#23287)

(cherry picked from commit f61119c0f3caa5963f00fa158f9d27f8f8e2cda8)

dstandish committed:
Optimize 2.3.0 pre-upgrade check queries (#23458)

We have to check for rows that are missing either corresponding TI or DR and move them out of table before adding FKs.  We were doing correlation in the JOIN condition but it appears postgres does *not* like this so here we move correlation to WHERE.

(cherry picked from commit 8e3d6c30e2409f56f9e4a55fe16769d23f5e3012)

jedcunningham committed:
Docs: Python 3.10 is now supported (#23457)

(cherry picked from commit 04eb78bc0da94acd898b88dc00145b20b748a8c0)

potiuk committed:
Unify approach for user questions asked in Breeze (#23335)

This change documents and unifies the approach we've taken for
the user inut handling when it comes to confirmation questions.

(cherry picked from commit ebc1f14db3a1b14f2535462e97a6407f48b19f7c)

uranusjr committed:
Show warning if '/' is used in a DAG run ID (#23106)

(cherry picked from commit 451c7cbc42a83a180c4362693508ed33dd1d1dab)

potiuk committed:
Move non-opencontainer labeling of the image to breeze from Dockerfile (#23379)

* Extract "extra" labeling of the image to breeze from Dockerfile

Fixes: #21046

* Add more ArtifictHub-specific labels

Co-authored-by: Kamil Breguła <kamilbregula@apache.org>
(cherry picked from commit 5b1ab96865a6e8f18784f88f88f8b17981450cc6)

mik-laj committed:
Bump pre-commit hook versions (#22887)

(cherry picked from commit 2d109401b3566aef613501691d18cf7e4c776cd2)

jedcunningham committed:
Add Python 3.10 trove classifier (#23464)

(cherry picked from commit 673f8940786b33663a5eb47047af2d7c2a14df99)

ephraimbuddy committed:
Remove remaining Python3.6 references (#23474)

(cherry picked from commit d9902958448b9d6e013f90f14d2d066f3121dcd5)

bbovenzi committed:
Remove color change for highly nested groups (#23482)

(cherry picked from commit 14befe3ad6a03f27e20357e9d4e69f99d19a06d1)

bbovenzi committed:
Visually distinguish task group summarys (#23488)

Bold task groups names and darken their bottom row border.

(cherry picked from commit 0fde90d92ae306f37041831f5514e9421eee676b)

jedcunningham committed:
Fix literal cross product expansion (#23434)

(cherry picked from commit 3fb8e0b0b4e8810bedece873949871a94dd7387a)

tanelk committed:
Use kubernetes queue in kubernetes hybrid executors (#23048)

When using "hybrid" executors (`CeleryKubernetesExecutor` or `LocalKubernetesExecutor`),
then the `clear_not_launched_queued_tasks` mechnism in the `KubernetesExecutor` can
reset the queued tasks, that were given to the other executor.

`KuberneterExecutor` should limit itself to the configured queue when working in the
"hybrid" mode.

(cherry picked from commit ae19eab3b4af98756200843805be882ce02a7d08)

edithturn committed:
Changed word 'the' instead 'his' (#23493)

(cherry picked from commit ed0f8a745e8dbe25ac45318c0c0d5f5ccda67356)

leahecole committed:
Replace DummyOperator references in docs (#23502)

(cherry picked from commit 69e361ccc08f9e72be5a6e39bfd5be8e4ee6387d)

bbovenzi committed:
Expand/collapse all groups (#23487)

* Add expand/collapse all groups button to Grid

* add tests

* add comments

* Switch to 2 icon buttons

Disable buttons if all groups are expanded or collapsed

* Update localStorage key

(cherry picked from commit 83784d9e7b79d2400307454ccafdacddaee16769)

joppevos committed:
Move tests command in new breeze (#23445)

(cherry picked from commit 7ba4e35a9d1b65b4c1a318ba4abdf521f98421a2)

dstandish committed:
Only count bad refs when `moved` table exists (#23491)

This keeps the logic to fail without upgrading when (A) there are bad rows and
(B) the "moved" table already exists. But we optimize so that we don't count
the bad rows unless the "moved" table is there. Previously we counted always,
but the first time a user attempts upgrade, the tables won't be there so
there's no point in counting.

Instead what we do is skip right to the CTAS, creating the _airflow_moved
tables. If there aren't any rows in the "moved" table, then we delete the table
immediately.

Also included here is a delete optimization, where we join to the moved table
instead of running the not exists query again.

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
Co-authored-by: Ash Berlin-Taylor <ash@apache.org>
(cherry picked from commit 6cc41abf6912fd2705b9ef7cf368c888c43c8af8)

ashb committed:
Change approach to finding bad rows to LEFT OUTER JOIN. (#23528)

Rather than sub-selects (two for count, or one for the CREATE TABLE).

For a _large_ database (27m TaskInstances, 2m DagRuns) this takes the
time from 10minutes to around 3 minutes per table (we have 3) down to 3
minutes per table. (All times on Postgres.)

Before:

```sql
CREATE TABLE _airflow_moved__2_3__dangling__rendered_task_instance_fields AS
SELECT
  rendered_task_instance_fields.dag_id AS dag_id,
  rendered_task_instance_fields.task_id AS task_id,
  rendered_task_instance_fields.execution_date AS execution_date,
  rendered_task_instance_fields.rendered_fields AS rendered_fields,
  rendered_task_instance_fields.k8s_pod_yaml AS k8s_pod_yaml +
FROM
  rendered_task_instance_fields
WHERE
  NOT (
    EXISTS (
      SELECT
        1
      FROM
        task_instance
        JOIN dag_run ON dag_run.dag_id = task_instance.dag_id
        AND dag_run.run_id = task_instance.run_id
      WHERE
        rendered_task_instance_fields.dag_id = task_instance.dag_id
        AND rendered_task_instance_fields.task_id = task_instance.task_id
        AND rendered_task_instance_fields.execution_date = dag_run.execution_date
    )
  )
```

After:

```sql
CREATE TABLE _airflow_moved__2_3__dangling__rendered_task_instance_fields AS
SELECT
  rendered_task_instance_fields.dag_id AS dag_id,
  rendered_task_instance_fields.task_id AS task_id,
  rendered_task_instance_fields.execution_date AS execution_date,
  rendered_task_instance_fields.rendered_fields AS rendered_fields,
  rendered_task_instance_fields.k8s_pod_yaml AS k8s_pod_yaml +
FROM
  rendered_task_instance_fields
  LEFT OUTER JOIN dag_run ON rendered_task_instance_fields.dag_id = dag_run.dag_id
  AND rendered_task_instance_fields.execution_date = dag_run.execution_date
  LEFT OUTER JOIN task_instance ON dag_run.dag_id = task_instance.dag_id
  AND dag_run.run_id = task_instance.run_id
  AND rendered_task_instance_fields.task_id = task_instance.task_id
WHERE
  task_instance.dag_id IS NULL
  OR dag_run.dag_id IS NULL
;
```

(cherry picked from commit 22a9293ff8f48411d39074d9bc88af35abe9850f)

potiuk committed:
Seperate provider verification as standalone breeze command (#23454)

This is another step in simplifying and converting to Python all of
the CI/local development tooling.

This PR separates out verification of providers as a separate
breeze command `verify-provider-packages`. It was previously part of
"prepare_provider_packages.py" but it has been now
extracted to a separate in-container python file and it was
wrapped with breeze's `verify-provider-packages` command.

No longer provider verification is run with "preparing provider docs"
nor "preparing provider packages" - it's a standaline command.

This command is also used in CI now to run the tests:

* all provider packages are built and created on CI together with
  airflow version
* the packages are installed inside the CI image and providers are
  verified
* the 2.1 version of Airflow is installed together with all 2.1
  - compatible providers and provider verification is run there too.

This all is much simpler now - we got rediof some 500 lines of bash
code again in favour of breeze python code.

Fixes: #23430
(cherry picked from commit 3ed07474649b1e202f9b106105fef21f7b2cfddc)

potiuk committed:
Fix accidental including of providers in airflow package (#23552)

The change #23454 accidentally remove INSTALL_PROVIDERS_FROM_SOURCES
setting to "false" which resulted in airflow package containing all
providers. This has been caught by our tests (but it was only
visible after merging)

This PR brings the variable back.

(cherry picked from commit 29ec8d0a06f0f548083bc4f31d0bfe61286a3342)

potiuk committed:
Add logging in to Github Registry for breeze pull (#23551)

All of the Airlfow Images are Public in ghcr.io but default setting
for iamges is "private" and when users want to build CI workflows
in their forks, had to manually change their images to Public, so
that ci.yml workflow can pull the images prepared in the build-images
workflow.

This PR adds logging in for `breeze pull` command when GITHUB_TOKEN
is available, also the workflow gets packages: read permissions.

This way ci should works in forks of users without any action from
user except first-time workflow enabling.

(cherry picked from commit 953b85d8a911301c040a3467ab2a1ba2b6d37cd7)

potiuk committed:
Add IPV6 form of the address in cassandra status check (#23537)

This PR fixes problem introduced in 3.0.26 of cassandra image which
adds square brackets around IP address regardless of its type.

The problem was workarounded by pinning cassandra to 3.0.25 in
the ##23522 as a quick fix, but this one introducec permanent,
future-proof solution.

Based on discussion in https://issues.apache.org/jira/browse/CASSANDRA-17612

Fixes: #23523
(cherry picked from commit 22a564296be1aee62d738105859bd94003ad9afc)

potiuk committed:
Refactor Breeze to group related methods and classes together (#23556)

This change refactors Breeze classes to more consistent approach.

* The "commands" package only contains commands
* All Parameters (BuildCi, BuildProd, BuildDoc, Shell) are now
  in "params" package
* Required/Optional Build args are now members of the
  BuildCiParams, BuildProdParams which makes the params
  much more self-contained..
* All utils are in "utils" package

This helps with avoiding circular imports (all utios are now
standalone and do not use any of the commands.

Co-authored-by: eladkal <45845474+eladkal@users.noreply.github.com>
(cherry picked from commit 35b917396fa35e70c47e114da8f178b1afeca5bc)

potiuk committed:
Fix _PIP_ADDITIONAL_REQUIREMENTS case for docker-compose (#23517)

Recent versions of Airflow do not allow to run `pip install` as
root but the `init` job runs as root so when the variable
_PIP_ADDITIONAL_REQUIREMENTS is set, the init container fails.

This PR forces _PIP_ADDITIONAL_REQUIREMENTS to be empty for the init
job.

(cherry picked from commit efed15bb6de8ac214b7029659ea08ce079116274)

potiuk committed:
Add slim images to release process (#23391)

This PR adds slim images to release process of Airflow.

Those images are small as they do not contain any extras.

Fixes: #20849
(cherry picked from commit 6e5955831672c71bfc0424dd50c8e72f6fd5b2a7)

yeachan153 committed:
Move dag_processing.processor_timeouts to counters section (#23393)

(cherry picked from commit fcfaa8307ac410283f1270a0df9e557570e5ffd3)

mthakare-onshape committed:
Update dags.rst (#23579)

Update missing bracket

(cherry picked from commit 827bfda59b7a0db6ada697ccd01c739d37430b9a)

Commit:
Fix dag-processor fetch metabase config (#23575)

(cherry picked from commit 9837e6d813744e3c5861c32e87b3aeb496d0f88d)

potiuk committed:
tHe output of commands of Breeze are only generated when they change (#23570)

Previously we generated output of all the commands from Breeze always,
hoping that they will be the same, but rich already had two changes
in the format of the SVG files which made the output different and
breaking our PRs.

Temporarily we pinned rich to fix the output, but better solution is
to get the hash of all the configuration options and see if it changed,
and only run generation when it did. This way we keep automated
generation on pre-commit but we are protected from accidental change
of the output.

We also remove the rich limits and regenerated all svg files to ones
generated by 12.4.0. Also found a way to run the check if we should
run generation at all in pre-commit without prior installing breeze.

Fixes: #22908
(cherry picked from commit 7c7b001a6399f7e6aa0f08226749ad2f8d2ce3bc)

ephraimbuddy committed:
Fix scheduler crash when expanding with mapped task that returned none (#23486)

When task is expanded from a mapped task that returned no value, it
crashes the scheduler. This PR fixes it by first checking if there's
a return value from the mapped task, if no returned value, then error
in the task itself instead of crashing the scheduler

(cherry picked from commit 7813f996ab79b9e6ef07090194f1e621e4f90e17)

tanelk committed:
Pools with negative open slots should not block other pools (#23143)

(cherry picked from commit 7132be2f11db24161940f57613874b4af86369c7)

eladkal committed:
Fix `PythonVirtualenvOperator` templated_fields (#23559)

* Fix `PythonVirtualenvOperator` templated_fields
The `PythonVirtualenvOperator` templated_fields override `PythonOperator` templated_fields which caused functionality not to work as expected.
fixes: https://github.com/apache/airflow/issues/23557

(cherry picked from commit 1657bd2827a3299a91ae0abbbfe4f6b80bd4cdc0)

repl-chris committed:
Fix broken dagrun links when many runs start at the same time (#23462)

* Load requested dagrun even when there are many dagruns at (almost) the same time

* Fix code formatting issues

(cherry picked from commit 828016747ac06f6fb2564c07bb8be92246f42567)

josh-fell committed:
Clean up in-line f-string concatenation (#23591)

(cherry picked from commit 428a43995390b3623a51aa7bac7e21da69a8db22)

mpeteuil committed:
Apply specific ID collation to root_dag_id too (#23536)

In certain databases there is a need to set the collation for ID fields
like dag_id or task_id to something different than the database default.
This is because in MySQL with utf8mb4 the index size becomes too big for
the MySQL limits. In past pull requests this was handled
[#7570](https://github.com/apache/airflow/pull/7570),
[#17729](https://github.com/apache/airflow/pull/17729), but the
root_dag_id field on the dag model was missed. Since this field is used
to join with the dag_id in various other models ([and
self-referentially](https://github.com/apache/airflow/blob/451c7cbc42a83a180c4362693508ed33dd1d1dab/airflow/models/dag.py#L2766)),
it also needs to have the same collation as other ID fields.

This can be seen by running `airflow db reset` before and after applying
this change while also specifying `sql_engine_collation_for_ids` in the
configuration.

Other related PRs
[#19408](https://github.com/apache/airflow/pull/19408)

(cherry picked from commit b7f862760ce4e20a284bb1933c9078e54851c518)

pierrejeambrun committed:
19943 Grid view status filters (#23392)

* Move tree filtering inside react and add some filters

* Move filters from context to utils

* Fix tests for useTreeData

* Fix last tests.

* Add tests for useFilters

* Refact to use existing SimpleStatus component

* Additional fix after rebase.

* Update following bbovenzi code review

* Update following code review

* Fix tests.

* Fix page flickering issues from react-query

* Fix side panel and small changes.

* Use default_dag_run_display_number in the filter options

* Handle timezone

* Fix flaky test

Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>
(cherry picked from commit 46c1c006f747798f7c80f615d791553f25dd1770)

potiuk committed:
Improve caching for multi-platform images. (#23562)

This is another attempt to improve caching performance for
multi-platform images as the previous ones were undermined by a
bug in buildx multiplatform cache-to implementattion that caused
the image cache to be overwritten between platforms,
when multiple images were build.

The bug is created for the buildx behaviour at
https://github.com/docker/buildx/issues/1044 and until it is fixed
we have to prpare separate caches for each platform and push them
to separate tags.

That adds a bit overhead on the building step, but for now it is
the simplest way we can workaround the bug if we do not want to
manually manipulate manifests and images.

(cherry picked from commit 9a6baab5a271b28b6b3cbf96ffa151ac7dc79013)

edithturn committed:
Use inclusive words in apache airflow project (#23090)

(cherry picked from commit d7b85d9a0a09fd7b287ec928d3b68c38481b0225)

potiuk committed:
Update min requirements for rich to 12.4.1 (#23604)

(cherry picked from commit 1bd75ddbe3b1e590e38d735757d99b43db1725d6)

potiuk committed:
Make Breeze help generation indepdent from having breeze installed (#23612)

Generation of Breeze help requires breeze to be installed. However
if you have locally installed breeze with different dependencies
and did not run self-upgrade, the results of generation of the
images might be different (for example when different rich
version is used). This change works in the way that:
* you do not have to have breeze installed at all to make it work
* it always upgrades to latest breeze when it is not installed
* but this only happens when you actually modified some breeze code

(cherry picked from commit e36868ba1b8a26f838851d1d09cca9db64cbcdc8)

potiuk committed:
Add slim images to docker-stack docs index (#23601)

(cherry picked from commit b0406f58f0c51db46d2da7c7c84a0b5c3d4f09ae)

mhenc committed:
Implement send_callback method for CeleryKubernetesExecutor and LocalKubernetesExecutor (#23617)

(cherry picked from commit c5b72bf30c8b80b6c022055834fc7272a1a44526)

jakubno committed:
Fix: Exception when parsing log #20966 (#23301)

* UnicodeDecodeError: 'utf-8' codec can't decode byte 0xXX in position X: invalid start byte

  File "/opt/work/python395/lib/python3.9/site-packages/airflow/hooks/subprocess.py", line 89, in run_command
    line = raw_line.decode(output_encoding).rstrip()            # raw_line ==  b'\x00\x00\x00\x11\xa9\x01\n'
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa9 in position 4: invalid start byte

* Update subprocess.py

* Update subprocess.py

* Fix:  Exception when parsing log #20966

* Fix:  Exception when parsing log #20966

 Another alternative is: try-catch it.

e.g.

```
            line = ''
            for raw_line in iter(self.sub_process.stdout.readline, b''):
                try:
                    line = raw_line.decode(output_encoding).rstrip()
                except UnicodeDecodeError as err:
                    print(err, output_encoding, raw_line)
                self.log.info("%s", line)
```

* Create test_subprocess.sh

* Update test_subprocess.py

* Added shell directive and license to test_subprocess.sh

* Distinguish between raw and decoded lines as suggested by @uranusjr

* simplify test

Co-authored-by: muhua <microhuang@live.com>
(cherry picked from commit 863b2576423e1a7933750b297a9b4518ae598db9)

ecerulm committed:
Prevent KubernetesJobWatcher getting stuck on resource too old (#23521)

* Prevent KubernetesJobWatcher getting stuck on resource too old

If the watch fails because "resource too old" the
KubernetesJobWatcher should not retry with the same resource version
as that will end up in loop where there is no progress.

* Reset ResourceVersion().resource_version to 0

(cherry picked from commit dee05b2ebca6ab66f1b447837e11fe204f98b2df)

raphaelauv committed:
[FEATURE] update K8S-KIND to 0.13.0 (#23636)

(cherry picked from commit 9b8cffe1c499833be758fbdbc7f271e19cad410f)

raphaelauv committed:
[FEATURE] add K8S 1.24 support (#23637)

(cherry picked from commit 8116bc915c6a7698ed64b63a72d8170016e35dce)

humit0 committed:
Fix typo issue (#23633)

(cherry picked from commit f313e144f751cc52ea144bf9dd99b1c85f15d469)

gmsantos committed:
Add index for event column in log table (#23625)

(cherry picked from commit 5d8cda8c5be42c8daaaa904d29a1011833c0c699)

dstandish committed:
Don't run pre-migration checks for downgrade (#23634)

These checks are only make sense for upgrades.  Generally they exist to resolve referential integrity issues etc before adding constraints.  In the downgrade context, we generally only remove constraints, so it's a non-issue.

(cherry picked from commit 9ab9cd47cff5292c3ad602762ae3e371c992ea92)

subkanthi committed:
Added postgres 14 to support versions(including breeze) (#23506)

* Added postgres 14 to support versions(including breeze)

(cherry picked from commit 6169e0a69875fb5080e8d70cfd9d5e650a9d13ba)

ishiis committed:
remove `--` in `./breeze build-docs` command (#23671)

(cherry picked from commit 3fa57168a520d8afe0c06d8a0200dd3517f43078)

potiuk committed:
Synchronize support for Postgres and K8S in docs (#23673)

We just added support for Postgres 14 and K8S 1.24 and since we
did not have any changes to support either in main we are bringing
the support to 2.3 line as well.

This documentation syncs all remaining places where it should be
updated.

(cherry picked from commit ea15277563c472efe4ef2efaa2424d728a80f987)

potiuk committed:
Add wildcard possibility to `package-filter` parametere (#23672)

the glob parameters (for example `apache-airflow-providers-*`) did
not work because only fixed list of parameters was allowed.

This PR converts the package-filter parameter to stop verifying the
value passed - so autocomplete continues to work but you should
still be able to use glob.

It also removes few places where the parameters were used with
`--` separator.

potiuk committed:
Replace "absolute()" with "resolve()" in pathlib objects (#23675)

TIL that absolute() is an undocumented in Pathlib and that we
should use resolve() instead.

So this is it.

(cherry picked from commit 094e50eb439e1c22e14ee78f375b7e00c71e9269)

potiuk committed:
Upgrade `pip` to latest released 22.1.0 version (#23665)

We are finally able to get rid of the annoying false-positive
warnings and we have finally a chance on having warning-free
installation during docker builds.

dstandish committed:
Shorten max pre-commit hook name length (#23677)

When names are too long, pre-commit output looks very ugly and takes up 2x lines. Here I reduce max length just a little bit further so that pre-commit output renders properly on a macbook pro 16" with terminal window splitting screen horizontally.

(cherry picked from commit c74e67a4ffb94ef5152b34fa96e88f1cbfe0c75a)

bbovenzi committed:
Move around overflow, position and padding (#23044)

(cherry picked from commit 028087b5a6e94fd98542d0e681d947979eb1011f)

bbovenzi committed:
Fix expand/collapse all buttons (#23590)

* communicate via customevents

* Handle open group logic in wrapper

* fix tests

* Make grid action buttons sticky

* Add default toggle fn

* fix splitting task id by '.'

* fix missing dagrun ids

(cherry picked from commit afdfece9372fed83602d50e2eaa365597b7d0101)

potiuk committed:
Disable Flower by default from docker-compose (#23685)

(cherry picked from commit 6eda265a41fc7c48692305b27c64f811ff02fb94)

potiuk committed:
Fix property name in breeze Shell Params (#23696)

The rename from #23562 missed few shell_parms usage where it
also should be replaced.

(cherry picked from commit 4afa8e3cecf1e4a2863715d14a45160034ad31a6)

potiuk committed:
Clarify that bundle extras should not be used for PyPi installs (#23697)

The bundle extras we have are only used for development and they
should not be used to install airflow from PyPI. This update
to documentation clarifies it.

Closes: #23692
(cherry picked from commit 310002e44887847991b0864bbf9a921c7b11e930)

potiuk committed:
Add environment check and build image check for more Breeze commands (#23687)

Several commands of Breeze depends on docker, docker compose
being available as well as breeze image. They will work
fine if you "just" built the image but they might benefit
from the image being rebuilt (to make sure all latest
dependencies are installed in the image). The common checks
done in "shell" command for that are now extracted to common
utils and run as first thing in those commands that need it.

(cherry picked from commit 3f4ab6c5bbdbfe4cc14eff35da8c2370b137f0fd)

bbovenzi committed:
Add UI tests for /utils and /components (#23456)

* Add UI tests for /utils and /components

* add test for Table

* Address PR feedback

* Fix window prompt var

* Fix TaskName test from rebase

* fix lint errors

(cherry picked from commit 694e380e52a1b2c9ad74148e1dda2914a0496b70)

mik-laj committed:
Add slim image to docs/docker-stack/README.md (#23710)

mik-laj committed:
Use profiles to disable flower in docker-compose (#23709)

(cherry picked from commit de3c0389eb42fa59eb9a4ad9caa2cd85baa367a0)

pingzh committed:
Ensure execution_timeout as timedelta (#23655)

(cherry picked from commit 6f82fc70ec91b493924249f062306330ee929728)

tirkarthi committed:
Handle invalid date parsing in webserver views. (#23161)

* Handle invalid date from query parameters in views.

* Add tests.

* Use common parsing helper.

* Add type hint.

* Remove unwanted error check.

* Fix extra_links endpoint.

(cherry picked from commit 9e25bc211f6f7bba1aff133d21fe3865dabda53d)

bbovenzi committed:
Fix grid details header text overlap (#23728)

Move top margin to each breadcrumb component to make sure that there is no overlap when the header wraps with long names.

(cherry picked from commit f77a6917264c0a2261e9d78fcfdc31a6e5f9760f)

dstandish committed:
Add version to migration prefix (#23564)

We don't really need the alembic revision id in the filename.  having version instead is much more useful.  having both of them takes up too much space.

(cherry picked from commit 741f8029115cc4698f9451b62546c3b1ff3d3a9c)

potiuk committed:
Add typing for airflow/configuration.py (#23716)

* Add typing for airflow/configuration.py

The configuraiton.py did not have typing information and it made
it rather difficult to reason about it-especially that it went
a few changes in the past that made it rather complex to
understand.

This PR adds typing information all over the configuration file

(cherry picked from commit 71e4deb1b093b7ad9320eb5eb34eca8ea440a238)

bbovenzi committed:
Remove titles from link buttons (#23736)

(cherry picked from commit 239a9dce5b97d45620862b42fd9018fdc9d6d505)

davidcaron committed:
Add 'reschedule' to the serialized fields for the BaseSensorOperator (#23674)

fix #23411

(cherry picked from commit f9e2a3051cd3a5b6fcf33bca4c929d220cf5661e)

potiuk committed:
Increase timeout for Helm Chart executor upgrade tests (#23759)

(cherry picked from commit 9dcca989ab6e2671c7278aa39fda0c4443232049)

pingzh committed:
Fix task log is not captured (#23684)

when StandardTaskRunner runs tasks with exec

Issue: https://github.com/apache/airflow/issues/23540
(cherry picked from commit e453e68158b65177572a3a1f95d8542c3a38d4e7)

jbampton committed:
Change `Github` to `GitHub` (#23764)

(cherry picked from commit 4a85370ca997015b75660192d2a481f09734c99d)

potiuk committed:
Add tagging image as latest for CI image wait (#23775)

The "wait for image" step lacked --tag-as-latest which made the
subsequent "fix-ownership" step run sometimes far longer than
needed - because it rebuilt the image for fix-ownership case.

Also the "fix-ownership" command has been changed to just pull
the image if one is missing locally rather than build. This
command might be run in an environment where the image is missing
or any other image was build (for example in jobs where an image
was build for different Python version) in this case the command
will simply use whatever Python version is available (it does
not matter), or in case no image is available, it will pull the image
as the last resort.

(cherry picked from commit 5e3f652397005c5fac6c6b0099de345b5c39148d)

jedcunningham committed:
Fix auto upstream dep when expanding non-templated field (#23771)

If you tried to expand via xcom into a non-templated field without
explicitly setting the upstream task dependency, the scheduler would
crash because the upstream task dependency wasn't being set
automatically. It was being set only for templated fields, but now we do
it for both.

(cherry picked from commit 3849ebb8d22bbc229d464c4171c9b5ff960cd089)

potiuk committed:
Fix python version used for cache preparaation (#23785)

Cache preparation on CI used default (Python 3.7) version of the
image. It had an influence on time of "full build needed" only and
for users who wanted to build breeze image for Python version
different than default Python 3.7.

It had no big influence on "main" builds" because in main we are
build images with "upgrade-to-newer-dependencies" which takes
longer anyway.

(cherry picked from commit 46abf376fbecc59b37db674fdf2a865d4c28a722)

potiuk committed:
Further speed up fixing ownership in CI (#23782)

After #23775 I noticed that there is yet another small improvement
area in the CI buld speed. Currently build-ci-image builds and push
only "commit-tagged" images, but "fix-ownership" requires
the "latest" image to run.

This PR adds --tag-as-latest option also to build-image and
build-prod-image commands - similarly as for the pull-image and
pull-prod-image. This will retag the "commit" images as latest in the
build-ci-images step and allow to save 1m on pulling the latest image
before fix-ownership (bringing it back to 1s overhead)

(cherry picked from commit 252ef66438ecda87a8aac4beed1f689f14ee8bec)

jianyuan committed:
Modify db clean to also catch the ProgrammingError exception (#23699)

(cherry picked from commit a80b2fcaea984813995d4a2610987a1c9068fdb5)

ephraimbuddy committed:
Update version to 2.3.1 and add release notes

potiuk committed:
Fix python version command (#23818)

(cherry picked from commit 8f3ce33e921056db3bac34dcf38b238cbf417279)

bbovenzi committed:
Grid data: do not load all mapped instances (#23813)

* only get necessary task instances

* add comment

* encode_ti -> get_task_summary

(cherry picked from commit 7ab5ea7853df9d99f6da3ab804ffe085378fbd8a)

ianbuss committed:
Fix regression in ignoring symlinks (#23535)

(cherry picked from commit 8494fc7036c33683af06a0e57474b8a6157fda05)

ianbuss committed:
Fix provider import error matching (#23825)

(cherry picked from commit 6f5749c0d04bd732b320fcbe7713f2611e3d3629)

tirkarthi committed:
Fix secrets rendered in UI when task is not executed. (#22754)

(cherry picked from commit ce8ea6691820140a0e2d9a5dad5254bc05a5a270)

potiuk committed:
Fix retrieval of deprecated non-config values (#23723)

It turned out that deprecation of config values did not work as
intended. While deprecation worked fine when the value was specified
in configuration value it did not work when `run_as_user` was used.

In those cases the "as_dict" option was used to generate temporary
configuratin and this temporary configuration contained default value
for the new configuration value - for example it caused that
the generated temporary value contained:

```
[database]
sql_alchemy_conn=sqlite:///{AIRFLOW_HOME}/airflow.db
```

Even if the deprecated `core/sql_alchemy_conn` was set (and no
new `database/sql_alchemy_conn` was set at the same time.

This effectively rendered the old installation that did not convert
to the new "database" configuration not working for run_as_user, because
the tasks run with "run_as_user" used wrong, empty sqlite database
instaead of the one configured for Airflow.

Also during adding tests, it turned out that the mechanism was also
not working as intended before - in case `_CMD` or `_SECRET` were used
as environment variables rather than configuration. In those cases
both _CMD and _SECRET should be evaluated during as_dict() evaluation,
because the "run_as_user" might have not enough permission to run the
command or retrieve secret. The _cmd and _secret variables were only
evaluated during as_dict() when they were in the config file (note
that this only happens when include_cmd, include_env, include_secret
are set to True).

The changes implemented in this PR fix both problems:

* the _CMD and _SECRET env vars are evaluated during as_dict when the
  respective include_* is set
* the defaults are only set for the values that have deprecations
  in case the deprecations have no values set in either of the ways:
    * in config file
    * in env variable
    * in _cmd (via config file or env variable)
    * in _secret (via config file or env variable)

Fixes: #23679
(cherry picked from commit 888bc2e233b1672a61433929e26b82210796fd71)

repl-chris committed:
Automatically reschedule stalled queued tasks in CeleryExecutor (v2) (#23690)

Celery can lose tasks on worker shutdown, causing airflow to just wait on them
indefinitely (may be related to celery/celery#7266). This PR expands the
"stalled tasks" functionality which is already in place for adopted tasks, and
adds the ability to apply it to all tasks such that these lost/hung tasks can
be automatically recovered and queued up again.

(cherry picked from commit baae70c88ed45d4b45e64754cb3decb99472c601)

jedcunningham committed:
Document fix for broken elasticsearch logs with 2.3.0+ upgrade (#23821)

In certain upgrade paths, Airflow isn't given an opportunity to track
the old `log_id_template`, so document the fix for folks who run into
trouble.

(cherry picked from commit 6f738d93397a7b1429efb934c0ec2775c69f0738)

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,93 @@
+^^^^^^^^^^^^^^^^
+
+- Add slim images to docker-stack docs index (#23601)
+- Update dags.rst (#23579)
```
```suggestion
```

Neither of these need to be here.

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,93 @@
+- update K8S-KIND to 0.13.0 (#23636)
+- Refactor code references from tree to grid (#23254)
+- [FIX] remove python 3.6 (#23409)
+- Fix empty image preparation (#23304)
```
```suggestion
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,93 @@
+- Disable Flower by default from docker-compose (#23685)
+- Added postgres 14 to support versions(including breeze) (#23506)
+- add K8S 1.24 support (#23637)
+- update K8S-KIND to 0.13.0 (#23636)
```
```suggestion
```

Doesn't impact anything user facing.

potiuk committed:
Add missing "airflow-constraints-reference" parameter (#23844)

The build commands were missing "airflow-constraints-reference"
parameter and it always defaulted to constraints-main

(cherry picked from commit cc3ab432e32485e491d81c180ec807b275a0cafa)

potiuk committed:
Better fix for constraint-reference (#23845)

The previous fix (#23844) broke main on package verification
as the package verification used the same parameter that was set to
empty.

This change rmeoves some remnant from the "bash" version where
we had to check if variable was empty and also making the "constraint"
parameters accepting default values from the current branch to be used
also for build commands.

(cherry picked from commit 10a0d8e7085f018b7328533030de76b48de747e2)

potiuk committed:
Add limit for JPype1 (#23847)

The JPype1 limit has to be introduced because otherwise the 1.4.0
JPype1 breaks our ARM builds. The 1.4.0 did not release the sdist
version of the package. This made our cache refresh job to fail
as 1.4.0 version cannot be installed on ARM image.

The issue is captured in
https://github.com/jpype-project/jpype/issues/1069

(cherry picked from commit 3699be49b24ef5a0a8d8de81a149af2c5a7dc206)

ephraimbuddy committed:
Update release date on Release note
